### PR TITLE
mrazadar/PON-89/subscription-unit-tests

### PR DESCRIPTION
**Pull request by mrazadar**
This pull request includes various unit tests for the subscription feature. 

- added snapshot tests for `SubscriptionPage`
- added snapshot tests for `SubscriptionDetails`
- added component, snapshot tests for `SubscriptionCheckout`
- added snapshot tests for `SubscriptionContent`
- added snapshot tests for details components
- Increased coverage for `payment/utils`
- added unit tests for `subscription/details` redux
- added unit tests for `subscription/client-secret` redux

### CODE REVIEW

File: src/subscription/SubscriptionPage.messages.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import { defineMessages } from '@edx/frontend-platform/i18n';
+
+const messages = defineMessages({
+  'subscription.heading.page': {
```
Comment (julianajlk) [inline src/subscription/SubscriptionPage.messages.js:4]: This is repeated from L57 on `SubscriptionPage.jsx`, you don't need the translation with `id` and `defaultMessage` in both the `messages.js` file and in the `<FormattedMessage>` component

File: src/subscription/SubscriptionPage.messages.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    id: 'subscription.heading.page',
+    defaultMessage: 'Subscription',
+    description: 'Page heading for subscription.',
+  },
+  'subscription.loading.details': {
+    id: 'subscription.loading.details',
+    defaultMessage: 'Loading subscription...',
+    description: 'Message when subscription page is being loaded',
+  },
+  'subscription.loading.error': {
```
Comment (julianajlk) [inline src/subscription/SubscriptionPage.messages.js:14]: I don't see this being used? Is it in another commit?

File: src/subscription/__factories__/subscription.factory.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  .sequence('subscription_id')
+  .option('numProducts', 0)
+  .option('productType', 'Course Entitlement')
+  .option('numErrorMessages', 0)
+  .option('numInfoMessages', 0)
+  .attrs({
+    program_title: 'Blockchain Fundamentals',
+    organization: 'University of California, Berkeley',
+    certificate_type: 'verified',
+    // due_date: getDueDate(),
```
Comment (julianajlk) [inline src/subscription/__factories__/subscription.factory.js:36]: Let's not forget to discuss this in standup today!

File: src/subscription/checkout/SubscriptionCheckout.test.jsx [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    store = createStore(createRootReducer(), {}, applyMiddleware(thunkMiddleware));
+    // eslint-disable-next-line no-import-assign
+    analytics.sendTrackingLogEvent = jest.fn();
+    Cookies.result[process.env.CURRENCY_COOKIE_NAME] = undefined;
+
+    // Mock the response of loadStripe method
+    // const mockedStripe = {
+    //   elements: jest.fn(),
+    // };
+    // loadStripe.mockResolvedValue(mockedStripe);
```
Comment (julianajlk) [inline src/subscription/checkout/SubscriptionCheckout.test.jsx:119]: Intentional? Your tests still pass even without mocking stripe?

File: src/subscription/data/details/reducer.js [hunk 5]
```diff
@@ -18,6 +18,7 @@ const subscriptionDetailsInitialState = {
   certificateType: 'verified',
   organization: 'University of California, Berkeley',
   price: 49,
+  paymentMethod: 'stripe',
```
Comment (julianajlk) [inline src/subscription/data/details/reducer.js:21]: Do you know if we finalized the discussion, whether we'll get this from `subscription` BE or will we hard code it in the FE?

File: src/subscription/data/sagas.js [hunk 6]
```diff
@@ -10,7 +10,7 @@ import { fetchSubscriptionDetails, submitPayment } from './details/actions';
 import { handleFetchSubscriptionDetails, handleSubmitPayment } from './details/sagas';
 
 export default function* saga() {
-  yield takeEvery(fetchSubscriptionClientSecret.TRIGGER, handleFetchClientSecret);
-  yield takeEvery(fetchSubscriptionDetails.TRIGGER, handleFetchSubscriptionDetails);
-  yield takeEvery(submitPayment.TRIGGER, handleSubmitPayment);
+  yield takeLatest(submitPayment.TRIGGER, handleSubmitPayment);
```
Comment (julianajlk) [inline src/subscription/data/sagas.js:13]: Nice switch to `takeLatest()`, since we only need the latest value and not every single value

**Review by julianajlk — APPROVED** [review]
Appreciate the separate commits and the added tests for the previously written utils functions!

Just an FYI, we might not need to fetch the client secret ID anymore with Stripe's deferred payment intent, I am currently working on a proof-of-concept, I'll keep you updated.

File: src/subscription/__factories__/subscription.factory.js (hunk 3 above)
Comment (julianajlk) [reply src/subscription/__factories__/subscription.factory.js:36]: Update here is that we'll get this date from the `subscription` BE, not directly from Stripe from the MFE. BE will get it from Stripe.

File: src/subscription/SubscriptionPage.jsx [hunk 7]
```diff
@@ -28,14 +27,11 @@ import { ConfirmationModal } from './confirmation-modal/ConfirmationModal';
...
   const {
-    isEmpty,
     isRedirect,
-    summaryQuantity,
-    summarySubtotal,
   } = useSelector(subscriptionSelector);
   const intl = useIntl();
   const dispatch = useDispatch();
-  const [isPaymentSuccessfull] = useState(false);
+  const [isPaymentSuccessFull] = useState(false);
```
[the remark this answers is not in the document]
Comment (mrazadar) [reply src/subscription/SubscriptionPage.jsx]: updated to `isPaymentSuccessful`

File: src/subscription/SubscriptionPage.messages.js (hunk 2 above)
Comment (mrazadar) [reply src/subscription/SubscriptionPage.messages.js]: Removed

File: src/subscription/__factories__/subscription.factory.js (hunk 3 above)
Comment (mrazadar) [reply src/subscription/__factories__/subscription.factory.js:36]: I think we created a ticket for it to send it from BE?

File: src/subscription/data/sagas.js (hunk 6 above)
Comment (mrazadar) [reply src/subscription/data/sagas.js:13]: Thanks

File: src/subscription/__factories__/subscription.factory.js (hunk 3 above)
Comment (mrazadar) [reply src/subscription/__factories__/subscription.factory.js:36]: ah! you already ready replied, thanks 
> Update here is that we'll get this date from the subscription BE, not directly from Stripe from the MFE. BE will get it from Stripe.

**Review by MuhammadMattiullah — APPROVED** [review]

**Comment by mrazadar** [comment]
@julianajlk added replied to comments please review again. Also split the `PON-12` into two tickets for pre-post and post apis integration.

**Pull request merged**